controllers/etcd_controller.go

```diff
@@ -963,33 +961,53 @@ func (r *EtcdReconciler) updateEtcdErrorStatus(etcd *druidv1alpha1.Etcd, lastErr
 }
 
 func (r *EtcdReconciler) updateEtcdStatus(etcd *druidv1alpha1.Etcd, svc *corev1.Service, ss *appsv1.StatefulSet) error {
+
+	sts, err := r.waitUntilStatefulSetReady(ss)
+	if err != nil {
+		return err
+	}
+
 	svcName := svc.Name
 	etcd.Status.Etcd = druidv1alpha1.CrossVersionObjectReference{
-		APIVersion: ss.APIVersion,
-		Kind:       ss.Kind,
-		Name:       ss.Name,
+		APIVersion: sts.APIVersion,
+		Kind:       sts.Kind,
+		Name:       sts.Name,
 	}
 	conditions := []druidv1alpha1.Condition{}
-	for _, condition := range ss.Status.Conditions {
+	for _, condition := range sts.Status.Conditions {
 		conditions = append(conditions, convertConditionsToEtcd(&condition))
 	}
 	etcd.Status.Conditions = conditions
 
 	// To be changed once we have multiple replicas.
-	etcd.Status.CurrentReplicas = ss.Status.CurrentReplicas
-	etcd.Status.ReadyReplicas = ss.Status.ReadyReplicas
-	etcd.Status.UpdatedReplicas = ss.Status.UpdatedReplicas
-	etcd.Status.Ready = (health.CheckStatefulSet(ss) == nil)
+	etcd.Status.CurrentReplicas = sts.Status.CurrentReplicas
+	etcd.Status.ReadyReplicas = sts.Status.ReadyReplicas
+	etcd.Status.UpdatedReplicas = sts.Status.UpdatedReplicas
 	etcd.Status.ServiceName = &svcName
 	etcd.Status.LastError = nil
 	etcd.Status.ObservedGeneration = &etcd.Generation
+	etcd.Status.Ready = true
 
 	if err := r.Status().Update(context.TODO(), etcd); err != nil && !errors.IsNotFound(err) {
 		return err
 	}
 	return r.removeOperationAnnotation(etcd)
 }
 
+func (r *EtcdReconciler) waitUntilStatefulSetReady(sts *appsv1.StatefulSet) (*appsv1.StatefulSet, error) {
+	ss := &appsv1.StatefulSet{}
+	err := gardenerretry.UntilTimeout(context.TODO(), DefaultInterval, DefaultTimeout, func(ctx context.Context) (bool, error) {
+		if err := r.Get(context.TODO(), types.NamespacedName{Name: sts.Name, Namespace: sts.Namespace}, ss); err != nil {
+			return false, err
+		}
+		if err := health.CheckStatefulSet(ss); err != nil {
+			return false, err
```

Comment:
```suggestion
			return false, nil
```

Reply:
If we return an error here then the whole `wait` function doesn't make sense. Then we can just return the error to the reconcile function and backoff when the STS is not ready yet, and try again in the next reconciliation.

Reply:
Ahh sorry my bad. I thought returning `err` will end the retries.